lxd/network_zones.go

```diff
@@ -196,50 +208,69 @@ func networkZoneAccessHandler(entitlement auth.Entitlement) func(d *Daemon, r *h
 func networkZonesGet(d *Daemon, r *http.Request) response.Response {
 	s := d.State()
 
+	allProjects := shared.IsTrue(request.QueryParam(r, "all-projects"))
 	requestProjectName := request.ProjectParam(r)
 	effectiveProjectName, _, err := project.NetworkZoneProject(s.DB.Cluster, requestProjectName)
 	if err != nil {
 		return response.SmartError(err)
 	}
 
-	recursion := util.IsRecursionRequest(r)
+	if allProjects && request.QueryParam(r, "project") != "" {
+		return response.BadRequest(errors.New("Cannot specify a project when requesting all projects"))
+	}
 
-	var zoneNames []string
+	recursion := util.IsRecursionRequest(r)
```

Comment:
This code section is:
1. Performing a database call before performing a validation that can be performed prior to the call.
2. Parsing the request query parameter for "project" twice.
3. Getting the effective project (likely via a superfluous DB call) even when the request is not project specific.

Here's a suggested change which should do the same thing:
```golang
allProjects := shared.IsTrue(request.QueryParam(r, "all-projects"))
requestProjectName := request.QueryParam(r, "project")

if allProjects && requestProjectName != "" {
    return response.BadRequest(errors.New("Cannot specify a project when requesting all projects"))
}

if !allProjects {
    if requestProjectName == "" {
        requestProjectName = api.ProjectDefaultName
    }

    effectiveProjectName, _, err := project.NetworkZoneProject(s.DB.Cluster, requestProjectName)
    if err != nil {
        return response.SmartError(err)
    }

    // If the request is project specific, then set effective project name in the request context so that the authorizer can generate the correct URL.
    request.SetCtxValue(r, request.CtxEffectiveProjectName, effectiveProjectName)
}

recursion := util.IsRecursionRequest(r)
```

With this change, you can remove the setting of the effective project name in the request context inside the transaction (which also sticks out as a little odd).